File: pnmc-web/src/app/core/revision-de-campos/revision-de-campos.ts
/**
 * El vocabulario de la revisión campo por campo, en un solo sitio.
 *
 * <b>POR QUÉ VIVE EN `core` Y NO EN UNA DE LAS DOS PANTALLAS.</b> Lo usan las dos puntas de la
 * misma conversación: el panel institucional, que escribe las notas, y el panel de la organización,
 * que las atiende. Puesto en cualquiera de los dos, el otro tendría que importar de un `features/`
 * ajeno, que es la dependencia que el proyecto evita desde que `seccion-organizacion` tuvo que
 * mudar `NIVELES_DE_COBERTURA` a `panel-organizacion.api.ts`.
 *
 * <b>LOS NOMBRES SON LOS DEL CABLE.</b> `PNMC.Contracts/RevisionDeCamposContratos.cs`. Un campo
 * mal escrito aquí no falla al compilar: falla en pantalla como un hueco en blanco.
 */

/** Los tres estados de una revisión. Los mismos que `CK_RevisionesFestival_Estado`. */
export type EstadoDeLaRevision = 'borrador' | 'enviada' | 'cerrada';

/** Los dos estados de una nota. Los mismos que `CK_RevisionesFestivalObservaciones_Estado`. */
export type EstadoDeLaNota = 'pendiente' | 'atendida';

/**
 * Sobre qué se pide el cambio.
 *
 * NO ES DECORACIÓN: `principal` es la ficha del registro y `subregistro` una de sus realizaciones
 * —en un Festival, una fila de `dbo.VersionesFestival`—. «Correo de contacto» existe en las dos y
 * son dos columnas distintas.
 *
 * <b>ANTES SE LLAMABAN `festival` Y `perfil_versionado`.</b> Se renombraron el 17 de septiembre de
 * 2026, al pasar las revisiones de Festival a las tablas genéricas: son los mismos dos conceptos que
 * ya usaba Mercados, y con dos nombres distintos hacía falta traducir en medio para que dos módulos
 * dijeran lo mismo.
 */
export type AmbitoDelCampo = 'principal' | 'subregistro';

/** Un cambio pedido sobre un campo, tal como viaja en las dos direcciones. */
export interface ObservacionDeCampo {
  /** Cero mientras la nota solo existe en la pantalla del funcionario. */
  id: number;
  ambito: AmbitoDelCampo;
  subregistroId: number | null;
  seccionId: string;
  campoId: string;
  campoEtiqueta: string;
  valorObservado: string | null;
  nota: string;
  estado: EstadoDeLaNota;
  fechaAtencion: string | null;
}

/** El borrador entero: la observación general, las dos puntas y las notas. */
export interface RevisionDeCampos {
  id: number;
  /** El módulo del registro: `festivales`, `mercados`… */
  moduloId: string;
  registroId: string;
  registroNombre: string;
  estado: EstadoDeLaRevision;
  observacionGeneral: string | null;
  /** Quién envía. */
  revisorNombre: string | null;
  /** Quién recibe. */
  destinatarioNombre: string | null;
  organizacionNombre: string | null;
  fechaActualizacion: string | null;
  fechaEnvio: string | null;
  observaciones: ObservacionDeCampo[];
}

export interface CambioEntreEnviosFestival {
  campoId: string;
  seccionId: string;
  campoEtiqueta: string;
  valorAnterior: string | null;
  valorRecibido: string | null;
  respondeAjusteSugerido: boolean;
}

export interface ComparacionEnviosFestival {
  numeroEnvio: number;
  fechaEnvio: string;
  numeroEnvioAnterior: number | null;
  fechaEnvioAnterior: string | null;
  esPrimerEnvio: boolean;
  comparacionParcial: boolean;
  totalCampos: number;
  camposModificados: number;
  ajustesSugeridosAtendidos: number;
  otrosCambios: number;
  cambios: CambioEntreEnviosFestival[];
}

/** El cuerpo del PUT y del POST. Reemplaza la lista entera: lo que no venga se borra. */
export interface GuardarRevisionSolicitud {
  observacionGeneral: string | null;
  observaciones: {
    ambito: AmbitoDelCampo;
    subregistroId: number | null;
    seccionId: string;
    campoId: string;
    campoEtiqueta: string;
    valorObservado: string | null;
    nota: string;
  }[];
}

/** La carcasa que devuelve el servidor cuando el registro todavía no tiene revisión abierta. */
export function revisionVacia(registroId: string, registroNombre = '', moduloId = 'festivales'): RevisionDeCampos {
  return {
    id: 0,
    moduloId,
    registroId,
    registroNombre,
    estado: 'borrador',
    observacionGeneral: null,
    revisorNombre: null,
    destinatarioNombre: null,
    organizacionNombre: null,
    fechaActualizacion: null,
    fechaEnvio: null,
    observaciones: [],
  };
}

/**
 * El ámbito de un campo, deducido de su identificador.
 *
 * <b>ES UNA REGLA Y NO UNA CONVENCIÓN DE NOMBRES.</b> Todo identificador empieza por `festival.` o
 * por `perfilVersionado.`, y de ahí sale a qué tabla apunta la nota. Se deduce en vez de escribirse porque
 * son cuarenta y siete campos en la plantilla: un atributo más por campo es cuarenta y siete
 * ocasiones de poner el ámbito equivocado, y el equivocado NO falla —cuelga la nota de la tabla de
 * al lado, donde nadie la busca—.
 *
 * Un identificador sin prefijo conocido se trata como del Festival y no se inventa un tercero: es
 * el ámbito que no necesita un perfil versionado abierto, así que es el que menos daño hace si alguien se
 * equivoca al escribirlo.
 */
export function ambitoDelCampo(campoId: string): AmbitoDelCampo {
  return campoId.startsWith('perfilVersionado.') ? 'subregistro' : 'principal';
}

/**
 * La llave con la que se busca la nota de un campo.
 *
 * LLEVA EL PERFIL VERSIONADO DENTRO porque el mismo `campoId` se repite en cada perfil: la nota sobre el correo
 * de un perfil no es la de otro. Es la misma llave que impone
 * `UQ_RevisionesFestivalObservaciones_Campo` en la base.
 */
export function llaveDelCampo(campoId: string, subregistroId: number | null): string {
  return ambitoDelCampo(campoId) === 'subregistro'
    ? `subregistro|${subregistroId}|${campoId}`
    : `festival||${campoId}`;
}

File: pnmc-api/src/PNMC.Domain/Entities/RevisionDeCampos.cs
namespace PNMC.Domain.Entities;

/// <summary>
/// Una revisión institucional de CUALQUIER proceso del Ecosistema, con sus cambios campo por campo.
/// </summary>
/// <remarks>
/// <para>
/// <b>ES LA UNICA FORMA DEL CICLO DE REVISION, Y NO NOMBRA NINGUN PROCESO.</b> Identifica
/// el registro por <see cref="ModuloId"/> + <see cref="RegistroId"/>, el mismo par con el que lo
/// identifican la bitácora de auditoría y la procedencia. Nace para Mercados Musicales el 16 de
/// septiembre de 2026 y sirve tal cual para Escuelas, Escenarios, Redes y Lutería.
/// </para>
/// <para>
/// <b>LA FORMA ANTERIOR YA NO EXISTE.</b> Hubo una tabla por proceso —<c>RevisionesFestival</c>,
/// atada a <c>dbo.Festivales</c> con clave ajena obligatoria—, y durante un tiempo convivieron las
/// dos como deuda nombrada. El 18 de septiembre de 2026 se saldó: los expedientes se copiaron aquí,
/// las ediciones de festival dejaron de tener su propio par y las tablas viejas se retiraron en
/// <c>V20260918_01__el_ciclo_de_revision_en_una_sola_forma.sql</c>. Queda una sola forma.
/// </para>
/// <para>
/// El ciclo y la gobernanza son los mismos: <c>borrador</c> mientras el funcionario escribe,
/// <c>enviada</c> cuando la organización la recibe, <c>cerrada</c> cuando vuelve a enviar el
/// registro; y los dos nombres —quien envía y quien recibe— se copian al crear la fila.
/// </para>
/// </remarks>
public sealed class RevisionDeRegistroRow
{
    public long Id { get; set; }

    /// <summary>El módulo del registro: <c>mercados</c>, <c>festivales</c>, <c>escuelas</c>…</summary>
    public string ModuloId { get; set; } = string.Empty;

    /// <summary>El identificador del registro dentro de su módulo, como texto.</summary>
    public string RegistroId { get; set; } = string.Empty;

    /// <summary>borrador | enviada | cerrada.</summary>
    public string Estado { get; set; } = "borrador";

    public int IdUsuarioRevisor { get; set; }
    public string? RevisorNombre { get; set; }

    public int? IdUsuarioDestinatario { get; set; }
    public string? DestinatarioNombre { get; set; }

    public int? IdOrganizacion { get; set; }
    public string? OrganizacionNombre { get; set; }

    public string? ObservacionGeneral { get; set; }

    public DateTime FechaCreacion { get; set; }
    public DateTime? FechaActualizacion { get; set; }
    public DateTime? FechaEnvio { get; set; }
    public DateTime? FechaCierre { get; set; }
}

/// <summary>
/// Un cambio pedido sobre UN campo concreto de la ficha de un registro.
/// </summary>
/// <remarks>
/// <para>
/// <b>EL AMBITO NO ES DECORACION.</b> <c>principal</c> es la ficha del registro y
/// <c>subregistro</c> una de sus realizaciones —una edición—. «Lugar específico» existe en el
/// mercado y en cada una de sus ediciones: sin el ámbito, la nota no dice cuál corregir.
/// </para>
/// <para>
/// <b>EL ROTULO Y EL VALOR VIAJAN COPIADOS.</b> Cuando la organización corrija el campo, el valor de
/// hoy deja de existir —justo porque se pidió cambiarlo—, y un expediente que no dice sobre qué se
/// pidió el cambio no sirve de evidencia.
/// </para>
/// </remarks>
public sealed class RevisionDeRegistroObservacionRow
{
    public long Id { get; set; }
    public long IdRevision { get; set; }

    /// <summary>
    /// La revisión a la que pertenece.
    /// </summary>
    /// <remarks>
    /// EXISTE PARA PODER AÑADIR NOTAS A UNA REVISION QUE TODAVIA NO TIENE IDENTIFICADOR: en el
    /// primer guardado la fila y sus notas se escriben en la misma transacción.
    /// </remarks>
    public RevisionDeRegistroRow? Revision { get; set; }

    /// <summary>principal | subregistro.</summary>
    public string Ambito { get; set; } = "principal";

    /// <summary>La realización sobre la que se pide el cambio. Nulo cuando el ámbito es la ficha.</summary>
    public string? SubregistroId { get; set; }

    public string SeccionId { get; set; } = string.Empty;
    public string CampoId { get; set; } = string.Empty;
    public string CampoEtiqueta { get; set; } = string.Empty;
    public string? ValorObservado { get; set; }

    public string Nota { get; set; } = string.Empty;

    /// <summary>pendiente | atendida.</summary>
    public string Estado { get; set; } = "pendiente";

    public DateTime FechaCreacion { get; set; }
    public DateTime? FechaActualizacion { get; set; }
    public DateTime? FechaAtencion { get; set; }
    public int? IdUsuarioAtiende { get; set; }
}
